### PR TITLE
util/ranger: support use `like` to build range for new collation columns | tidb-test=pr/2247

### LINKED ISSUE #48181: Can't use `like` to build range when new collation is enabled
## Bug Report
After [pr 31278](https://github.com/pingcap/tidb/pull/31278) , we can't use `like` to build range when new collation is enabled. This doesn't make sense and will cause bad sql performance.
Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
CREATE TABLE `t1` (
  `id` varchar(40) COLLATE utf8mb4_general_ci NOT NULL,
  `name` varchar(100) COLLATE utf8mb4_general_ci DEFAULT NULL,
  `password` varchar(100) COLLATE utf8mb4_general_ci DEFAULT NULL,
  KEY `ix_name` (`name`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_general_ci;
```

### 2. What did you expect to see? (Required)
```
mysql> explain SELECT * FROM t1 cm use index(ix_name) WHERE cm.name LIKE  '18620391343%';
+-------------------------------+---------+-----------+--------------------------------------------------------+---------------------------------------------------------------------+
| id                            | estRows | task      | access object                                          | operator info                                                       |
+-------------------------------+---------+-----------+--------------------------------------------------------+---------------------------------------------------------------------+
| IndexLookUp_7                 | 250.00  | root      |                                                        |                                                                     |
| ├─IndexRangeScan_5(Build) | 250.00  | cop[tikv] | table:cm, index:ix_name(name) | range:["18620391343","18620391344"), keep order:false, stats:pseudo |
| └─TableRowIDScan_6(Probe) | 250.00  | cop[tikv] | table:cm                                               | keep order:false, stats:pseudo                                      |
+-------------------------------+---------+-----------+--------------------------------------------------------+---------------------------------------------------------------------+
3 rows in set (0.01 sec)
```

### 3. What did you see instead (Required)

```
mysql> explain SELECT * FROM t1 cm use index(ix_name) WHERE cm.name LIKE  '18620391343%';
+-------------------------------+----------+-----------+--------------------------------------------------------+---------------------------------------------------------+
| id                            | estRows  | task      | access object                                          | operator info                                           |
+-------------------------------+----------+-----------+--------------------------------------------------------+---------------------------------------------------------+
| IndexLookUp_8                 | 8000.00  | root      |                                                        |                                                         |
| ├─Selection_7(Build)      | 8000.00  | cop[tikv] |                                                        | like(test.company_member.logonname, "18620391343%", 92) |
| │ └─IndexFullScan_5     | 10000.00 | cop[tikv] | table:cm, index:ix_name(name) | keep order:false, stats:pseudo                          |
| └─TableRowIDScan_6(Probe) | 8000.00  | cop[tikv] | table:cm                                               | keep order:false, stats:pseudo                          |
+-------------------------------+----------+-----------+--------------------------------------------------------+---------------------------------------------------------+
4 rows in set (0.01 sec)
```

### 4. What is your TiDB version? (Required)

v6.5.3

### LINKED ISSUE #49138: Query result may be wrong when several filter merged into a point range on a prefix column
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t(a varchar(20) collate utf8mb4_general_ci, index ia(a(3)));
insert into t value('aa'),('aab'),('aabb');
select * from t use index (ia) where a >= 'aabb' and a <= 'aabd';
```

### 2. What did you expect to see? (Required)

```
> select * from t use index (ia) where a >= 'aabb' and a <= 'aabd';
+------+
| a    |
+------+
| aabb |
+------+
1 row in set (0.001 sec)
```

### 3. What did you see instead (Required)

```
> select * from t use index (ia) where a >= 'aabb' and a <= 'aabd';
+------+
| a    |
+------+
| aab  |
+------+
1 row in set (0.007 sec)

> explain select * from t use index (ia) where a >= 'aabb' and a <= 'aabd';
+-------------------------------+---------+-----------+----------------------+-----------------------------------------------------+
| id                            | estRows | task      | access object        | operator info                                       |
+-------------------------------+---------+-----------+----------------------+-----------------------------------------------------+
| IndexLookUp_8                 | 250.00  | root      |                      |                                                     |
| ├─IndexRangeScan_5(Build)     | 312.50  | cop[tikv] | table:t, index:ia(a) | range:["aab","aab"], keep order:false, stats:pseudo |
| └─Selection_7(Probe)          | 250.00  | cop[tikv] |                      | eq(test.t.a, "aab")                                 |
|   └─TableRowIDScan_6          | 312.50  | cop[tikv] | table:t              | keep order:false, stats:pseudo                      |
+-------------------------------+---------+-----------+----------------------+-----------------------------------------------------+
4 rows in set (0.012 sec)
```

### 4. What is your TiDB version? (Required)

only affects nightly

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: close #48181 close #49138

### What is changed and how it works?

1. For the `builder` in `pkg/util/ranger/points.go`, we provide a new parameter `convertToSortKey bool`. If it's `true`, we can build ranges for new collation columns in `newBuildFromPatternLike`. The user of `builder` should specify if it should convert the string values to the sort key or keep the original value.
    - Here we use `KeyWithoutTrimRightSpace()` and then +1 to calculate the end key. The `KeyWithoutTrimRightSpace` prevents regression compared to the existing behavior for binary collations. For other details, please see the comments.
2. Add two functions `pointsConvertToSortKey`, and `pointConvertToSortKey` to do the conversion work.
    - Previously, the `convertPoint` happens before converting to the sort key. Specifically, it's called in `points2Ranges`, `appendPoints2Ranges`, and `points2TableRanges`, which are right after `build()`. Now we try to keep this logic unchanged. So we do a `convertPoint` inside these two functions before the conversion to the sort key.
    - To implement 2., we need to pass the `newTp` into the `builder`. Therefore I merged it with the previous `Collator` parameter for the `build()`.
3. The previous `convertPoint` is not removed because it's needed to handle non-string cases. In such a case, we need to make the existing `convertPoint` able to correctly handle the new sort key cases. Specifically, we need to use the binary collation instead of the collation of the column for the `FieldType` passed into `convertPoint`.
    - I also added a `convertStringFTToBinaryCollate()` to reduce some repeated code.
4. In `pkg/util/ranger/checker.go`, remove the limitation that we can't use `LIKE` function on new collation column to build ranges.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Support using Range Scan on non-binary collation columns for `LIKE` functions.
```

### CODE REVIEW

**Review by tangenta — APPROVED** [review]

File: pkg/util/ranger/points.go [hunk 1]
```diff
@@ -135,6 +136,47 @@ func rangePointEqualValueLess(a, b *point) bool {
...
+	sortKey := p.value.GetBytes()
+	if collate.NewCollationEnabled() {
+		if !trimTrailingSpace {
+			sortKey = collate.GetCollator(newTp.GetCollate()).KeyWithoutTrimRightSpace(string(hack.String(sortKey)))
+		} else {
+			sortKey = collate.GetCollator(newTp.GetCollate()).Key(string(hack.String(sortKey)))
+		}
+	}
+	return &point{value: types.NewBytesDatum(sortKey), excl: p.excl, start: p.start}, nil
+}
```
Comment (winoros) [inline pkg/util/ranger/points.go:178]: Or
```
if !collate.NewCollationEnabled {
	return inputP
}
...
```
will be better?

File: pkg/util/ranger/points.go [hunk 2]
```diff
@@ -135,6 +136,47 @@ func rangePointEqualValueLess(a, b *point) bool {
 	return a.excl && !b.excl
 }
 
+func pointsConvertToSortKey(sctx sessionctx.Context, inputPs []*point, newTp *types.FieldType) ([]*point, error) {
+	if newTp.EvalType() != types.ETString ||
+		newTp.GetType() == mysql.TypeEnum ||
+		newTp.GetType() == mysql.TypeSet {
+		return inputPs, nil
+	}
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply pkg/util/ranger/points.go:148]: Added some explanations.

File: pkg/util/ranger/points.go (hunk 1 above)
Comment (time-and-fate) [reply pkg/util/ranger/points.go]: Updated. Merged this check into previous `if` branch.

**Pull request merged**